### PR TITLE
Storage refactoring

### PR DESCRIPTION
**Pull request by speedymonster**

### CODE REVIEW

File: mods/storage/test/storage.unit.test.ts [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+describe("@fonos/storage", () => {
+  let storage: Storage;
+
+  before(async () => {
+    storage = new Storage();
+    sandbox.stub(FonosService.prototype, "init").returns();
+  });
+  afterEach(() => sandbox.restore());
+
+  it("Should reject because is a directory", () => {
```
Comment (wardner) [inline mods/storage/test/storage.unit.test.ts]: The `it`  message must start in lowercase

**Review by wardner — COMMENTED** [review]
Reviewed by Wardner

File: mods/storage/test/storage.unit.test.ts (hunk 1 above)
Comment (rihernandez) [reply mods/storage/test/storage.unit.test.ts]: I agree with @Wardner , after all everything looks well.

File: mods/storage/test/storage.unit.test.ts [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+    sandbox.stub(path, "basename").returns(fileNameReturn);
+    sandbox
+      .stub(fs, "statSync")
+      .returns({...Stats.prototype, size: objectSize});
+    sandbox.stub(FonosService.prototype, "getService").returns({
+      uploadObject: () => {
+        return {
+          sendMessage: () =>
+            Promise.resolve({
+              stream: {
```
Comment (psanders) [inline mods/storage/test/storage.unit.test.ts]: You are using this code several times. Consider refactoring.

File: mods/storage/src/constants.ts [hunk 3]
```diff
@@ -0,0 +1,7 @@
+export const constants = {
+  HIGH_WATER_MARK: 1 * 1024,
+  APP_BUCKET: "apps",
+  RECORDINGS_BUCKET: "recordings",
+  PUBLIC_BUCKET: "public",
+  REJECT_UPLOAD_DIRECTORY: "Uploading a directory is not supported"
```
Comment (psanders) [inline mods/storage/src/constants.ts]: This is a simple log comment. It should follow the same pattern we use elsewhere(not on a constant)

File: mods/storage/src/constants.ts [hunk 4]
```diff
@@ -0,0 +1,7 @@
+export const constants = {
```
Comment (psanders) [inline mods/storage/src/constants.ts:19]: Add copywrite notice the header.

File: mods/storage/src/storage.ts [hunk 5]
```diff
@@ -64,64 +84,81 @@ export default class Storage extends FonosService {
...
+  /**
+   * Upload an object to Fonos Object Storage subsystem.
+   *
+   * @param {UploadObjectRequest} request - Object with information about the origin and
+   * destination of an object
+   * @param {string} request.bucket - Bucket at the Storage system
+   * @param {string} request.dir - Directory on the Storage system where your objec will be uploaded
+   * @param {string} request.filename - Path to the object to be uploaded
+   * @return {Promise<UploadObjectResponse>} localy accessible URL to the object
+
```
Comment (psanders) [inline mods/storage/src/storage.ts]: Unnecessary [Enter]

File: mods/storage/src/storage.ts [hunk 6]
```diff
@@ -64,64 +84,81 @@ export default class Storage extends FonosService {
...
+   * @throws if the path does not exist or if is a directory
+   * @throws if the directory does not exist
+   * @example
+   *
+   * const request = {
+   *    filename: '/path/to/file',
```
Comment (psanders) [inline mods/storage/src/storage.ts]: Refactor to use a double quote on all strings.

File: mods/storage/src/storage.ts [hunk 7]
```diff
@@ -146,44 +183,43 @@ export default class Storage extends FonosService {
...
-          } else {
-            resolve({
-              url: res.getUrl()
-            });
-          }
-        });
-    }).catch((e) => {
-      throw e;
-    });
+  private async getObjectURLInternal(
```
Comment (psanders) [inline mods/storage/src/storage.ts]: This method doesn't seem to add any value. Consider unifying `getObjectURL` and `getObjectURLInternal`

File: mods/storage/src/storage.ts [hunk 8]
```diff
@@ -146,44 +183,43 @@ export default class Storage extends FonosService {
...
+      .getObjectURL()
+      .sendMessage(utils.getObjectServiceUtils(request));
+
+    return result.getUrl();
   }
 
   // Internal API
-  uploadObjectSync(request: UploadObjectRequest) {
+  /**
+   * Upload an object to Fonos Object Storage subsystem.
```
Comment (psanders) [inline mods/storage/src/storage.ts]: Consider enhancing comment to clarify that this is a synchronous version of another function.

File: mods/storage/src/storage.ts [hunk 9]
```diff
@@ -200,7 +236,29 @@ export default class Storage extends FonosService {
   }
 
   // Internal API
-  getObjectURLSync(request: GetObjectURLRequest) {
+  /**
+   * Get Object URL.
```
Comment (psanders) [inline mods/storage/src/storage.ts]: Consider enhancing comment to clarify that this is a synchronous version of another function.

File: mods/storage/src/types.ts [hunk 10]
```diff
@@ -0,0 +1,20 @@
+export interface UploadObjectRequest {
```
Comment (psanders) [inline mods/storage/src/types.ts:19]: Please add the copyright notice.

File: mods/storage/src/utils.ts [hunk 11]
```diff
@@ -0,0 +1,88 @@
+import fs from "fs";
```
Comment (psanders) [inline mods/storage/src/utils.ts:19]: Please add the copyright notice.

File: mods/storage/test/storage.unit.test.ts [hunk 12]
```diff
@@ -0,0 +1,261 @@
+import Storage from "../src/storage";
```
Comment (psanders) [inline mods/storage/test/storage.unit.test.ts:19]: Please add the copyright notice.

**Review by psanders — COMMENTED** [review]
No additional comments.

**Review by psanders — COMMENTED** [review]
No additional comments.

**Review by rihernandez — COMMENTED** [review]
good job!

**Comment by wardner** [comment]
Nice!

**Pull request merged**